Expand channels to allow them to contain environment variables (#156)
conda-incubator/conda-lock: docs/authenticated_channels.md

# Authentication for channels

Conda lock supports two kinds of credentials used for channels

## Token based

These are used by [anaconda.org](https://anaconda.org/), [Anaconda Enterprise](https://www.anaconda.com/products/enterprise),
[Anaconda Team Edition](https://www.anaconda.com/products/team) and [Quetz](https://github.com/mamba-org/quetz).

To pass one of these channels specify them in your source with an environment variable

Make sure this environment variable is not expanded (quote types matter).

```sh
--channel 'http://host.com/t/$MY_REPO_TOKEN/channel'
```

If you accidentally pass a channel url that contains a token like so

```sh
--channel "http://host.com/t/$MY_REPO_TOKEN/channel"
```

> Then conda lock will detect the environment variabl used (provided that the environment variable used ends in a sensible suffix (KEY, TOKEN, PASS, etc)).

Reviewer: This seems too magical. Is anything lost by not having this magic?

The _name_ of the environment variable will form part of your lock and you will have to have that SAME environment variable set if you wish to run the install

```sh
# retrieve secrets from some store
source $(./get-lockfile-env-vars-from-secret-store)
# use the secrets as part of the conda-lock invocation

> conda-lock install -n my-env-with-a-secret condae-lock.yml

Reviewer: ```suggestion
conda-lock install -n my-env-with-a-secret conda-lock.yml
```

```

## Simple Auth

> For other channels (such as those self-managed) you may be using a more standard username/password auth

Reviewer: ```suggestion
For other channels (such as those self-managed) basic auth is supported
```
